hive-metastore/src/main/java/org/apache/iceberg/hive/MetastoreUtil.java

```diff
@@ -72,6 +73,23 @@ public static void alterTable(
     env.putAll(extraEnv);
     env.put(StatsSetupConst.DO_NOT_UPDATE_STATS, StatsSetupConst.TRUE);
 
-    ALTER_TABLE.invoke(client, databaseName, tblName, table, new EnvironmentContext(env));
+    try {
+      ALTER_TABLE.invoke(client, databaseName, tblName, table, new EnvironmentContext(env));
+    } catch (RuntimeException e) {
```

Comment:
How widely is `alterTable` used?
Wouldn't it be better (more general solution) to expose the wrapped `MetaException` instead?

Reply:
It's only being used at two places:

1.  To [update](https://github.com/apache/iceberg/blob/12f6d0d0a35522915f5a8da73973403fe12cb01c/hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java#L112) the table .
2. To [rename](https://github.com/apache/iceberg/blob/12f6d0d0a35522915f5a8da73973403fe12cb01c/hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java#L245) the table .

Are you proposing to throw non RTE `MetaException` at alter table . something like 
```
catch (RuntimeException e) {
      if (e.getCause() instanceof InvalidOperationException
          && e.getCause().getMessage() != null
          && e.getCause()
              .getMessage()
              .contains(
                  String.format(
                      "new table %s.%s already exists", table.getDbName(), table.getTableName()))) {
        throw new MetaException(
            String.format("Table already exists: %s.%s", table.getDbName(), table.getTableName()));
      }
      throw e;
    }

Reply:
I proposing something like this:
```
catch (RuntimeException e) {
      if (e.getCause() instanceof MetaException) {
        throw (MetaException) e.getCause()
      } else {
        throw e;
      }
    }
```

Reply:
Any thoughts on the test case [testRenameTableDestinationTableAlreadyExists](https://github.com/apache/iceberg/blob/bc6b6e65ee221fc4c4f7f72d3585b768f7beeecb/hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveCatalog.java#L1188) since it expects error message as 

```
.isInstanceOf(AlreadyExistsException.class)
        .hasMessageContaining("Table already exists");

```
Currently we have override this as 
```
.isInstanceOf(RuntimeException.class)
        .hasMessageContaining("new table newdb.table_renamed already exists");

Reply:
`alterTable` should return the `MetaException` and where we call it, we should handle the `InvalidOperationException` with the message of `new table %s.%s already exists` and translate it to an `AlreadyExistsException`.

We should only consider moving the exception if we find that all of the places where we call `alterTable` we would like to throw the same `AlreadyExistsException`

Reply:
`InvalidOperationException`  still will be wrapped with `RuntimeException` . So basically we will have to catch `RuntimeException` and unwrap with `AlreadyExistsException` . 

Or Am I missing something here ?

Reply:
Why not unwrap RTE with both `MetaException` as well as `InvalidOperationException` and translate to `org.apache.hadoop.hive.metastore.api.AlreadyExistsException` . Since all the callers of `alterTable` is already handling `org.apache.hadoop.hive.metastore.api.AlreadyExistsException`. 
I am thinking something like this:
```
catch (RuntimeException e) {
      // MetaException would be wrapped into RuntimeException during reflection
      if (e.getCause() instanceof MetaException) {
        throw (MetaException) e.getCause();
      }
      // InvalidOperationException would be wrapped into RuntimeException during reflection, as
      // java.lang.RuntimeException:InvalidOperationException(message:new table <> already exists)
      if (e.getCause() instanceof InvalidOperationException
              && e.getCause().getMessage() != null
              && e.getCause()
              .getMessage()
              .contains(
                      String.format(
                              "new table %s.%s already exists", table.getDbName(), table.getTableName()))) {
        throw new AlreadyExistsException(
                String.format("Table already exists: %s.%s", table.getDbName(), table.getTableName()));
      }
      throw e;
    }
```

Reply:
In the `HiveTableOperations` we are accustomed to handle Hive `MetaExceptions`, so I think handling `InvalidOperationException` there would be "natural".

The `alterTable` wraps the exceptions to a `RuntimeException` because of the `DynMethod` calls. I think this is a mistake. We should behave the same way as the `metaClients.run(client -> client...)` like calls, we should throw the original `MetaException` - my opinion here could be changes if this becomes a breaking change somewhere, but if it does not break code for someone, then we should change it.

Reply:
I made a mistake in the above command, I thought, that the root exception is `MetaException`, but in reality it is `TException`.

So fixing this issue, my comment correctly is below:
-----
The `alterTable` wraps the exceptions to a `RuntimeException` because of the `DynMethod` calls. I think this is a mistake. We should behave the same way as the `metaClients.run(client -> client...)` like calls, we should throw the original `TException` - my opinion here could be changes if this becomes a breaking change somewhere, but if it does not break code for someone, then we should change it.